### ISSUE TITLE
Crash due to "Corrupt MCB chain"

### ISSUE THREAD

**Issue by darkstar** [dosbox-staging/dosbox-staging#396]
I can get DosBox-Staging to consistently crash by playing a bit of Zone66 followed by any other game (in my example, Keen4)

I'm not sure if the corruption is because of DosBox or the game (Zone66). In any case, debugging this issue could be improved (e.g. report the error when it happens, not when the next program is started, or even silently repair the memory chain after the last program/game has exited to the DOS prompt)

![image](https://user-images.githubusercontent.com/43795/82741782-fbc51680-9d55-11ea-8e0d-6a4cee360782.png)

**Comment by dreamer** [comment]
I think this is the second time I see this "Corrupt MCB chain" error… raising the priority a little and putting this bug in important bugs list.

**Comment by darkstar** [comment]
If you want, I can get you the two games I tested it with. Or if you have any hints on how to debug this further, let me know. I can also try custom builds (though I don't have my Visual Studio set up currently, so I cannot at the moment build it myself)

**Comment by dreamer** [comment]
@darkstar Thank you :) I have my plate quite full already with plenty of stuff, that needs fixing, so help with this will be appreciated.

I'll investigate when I will have some free time, but until then, if you have spare cycles to work on this, then some tips :) :

We do support builds using msys2 as well as Visual Studio, but using VS+vcpkg is way, way, way easier. If you want to go with msys2, then documentation in BUILD.md file should be helpful in setting it up.

It seems like this "controlled crash" was introduced quite a long time ago to address pre-existing problem: 5fef7164a

It would be a good idea to check if the same crash happens in DOSBox 0.74-3 - if the problem does not happen there, it means regression happened sometime between 2010 and 2015 - these kinds of bugs are generally easy to bisect on Linux using Git… (unfortunately bisecting old DOSBox code on Windows will be a problem - upstream had broken VS solution for years, and requires some hacks and out-of-tree SDL patches to build on Windows :/).

**Comment by GranMinigun** [comment]
> I think this is the second time I see this "Corrupt MCB chain" error…

#277

It should also happen in SVN, but not 0.74-3.

**Comment by darkstar** [comment]
I wrote a small test program (just for kicks) that repeatedly allocates, reallocates and frees small memory blocks using int 21h functions 48h, 49h and 4ah. I let it run for a while, trying to trigger the bug, but without success. So it looks like this is indeed a bug in the game, in that it corrupts the PSP memory chain. Maybe its DOS extended does something which works on real DOS, but not on DOSbox. This needs further debugging, I think.

I will try and get a Linux build environment up and running, so that I can bisect

**Comment by darkstar** [comment]
0.73 also crashes (this is before the check was implemented so it results in an infinite loop)

I was not able to compile older versions, probably because the compiler or autotools I have are too new.

**Comment by kcgen** [comment]
Some other program that corrupt the memory chain:
- X-COM: Terror from the Deep: https://github.com/joncampbell123/dosbox-x/issues/498
- Word Perfect for DOS 5.1: https://github.com/joncampbell123/dosbox-x/issues/638
- The Frotz infocom game interpreter: https://www.vogons.org/viewtopic.php?t=43440

Assuming this is the program's fault (akin to a segfault or bus-error), then there's not much we can do. But given we can detect it, I wonder if the chain can be 'repaired'?

**Comment by darkstar** [comment]
I wonder how this would work in actual MS-DOS... I played Zone66 a lot back in the 90's and don't remember any crashes after quitting. I remember I was usually able to start another game right after. So maybe something is indeed different between how MSDOS handles MCBs, and how DosBox handles them.

One possible way to fix it would be to simply "clean" and destroy all MCBs after a program quits regularly (i.e. not as a TSR).

**Comment by darkstar** [comment]
reading the reports mentioned above, another idea to further debug this would be to check the MCB chain every time int21h functions 48h, 49h or 4ah are called. That might help to catch the error earlier.

Also I wonder if the SDA size has anything to do with it, as reported in  the WP bug report

**Comment by dreamer** [comment]
While investigating a different issue I found a bit of info, that might be relevant here.

Is this problem reproducible after using `loadfix`? To test, just add `loadfix` command (without any parameters) before starting Zone66.

**Comment by darkstar** [comment]
It changes *something*, although not for the better. After using `loadfix` to reserve 64k and trying to start a game in Zone66, the game immediately crashes (it even crashes DosBox):
![image](https://user-images.githubusercontent.com/43795/85933534-5b27c080-b8d8-11ea-9b2e-20222ba4c8eb.png)

This is probably where it corrupts the MCB chain, although this time it looks like it corrupts something else instead..

Note that this was tested with the same version I used initially (from 2020-05-24)

**Comment by kcgen** [comment]
Zone 66 causes a bunch of these `illegal read` errors when forcing the dynamic core:

``` text
Illegal read from ef04ef67, CS:IP        8:    c99b
Illegal read from ef04ef68, CS:IP        8:    c99b
Illegal read from ef04ef69, CS:IP        8:    c99b
Illegal read from ef04ef6a, CS:IP        8:    c99b
Illegal read from ef04ef7a, CS:IP        8:    c997
Illegal read from ef04ef76, CS:IP        8:    c99b
```

Here's the full log:
[:spiral_notepad: 
 zone66-gus-illegal-reads.txt](https://github.com/dosbox-staging/dosbox-staging/files/4841468/zone66-gus-illegal-reads.txt) 55 KB.

**Comment by LowLevelMahn** [comment]
> It changes _something_, although not for the better. After using `loadfix` to reserve 64k and trying to start a game in Zone66, the game immediately crashes (it even crashes DosBox):

loadfix is a program (initially ) developed by microsoft to fix bugs in their
exepack unpacking algorithm: http://www.os2museum.com/wp/exepack-and-the-a20-gate/

using UNP: http://unp.bencastricum.nl/
or exepack: https://www.bamsoftware.com/software/exepack/

to get rid of the exe packing could maybe fix the problem or get a more clear view on the real error

**Comment by darkstar** [comment]
The error is not related to expack at all. If you hit the expack bug, the program won't load correctly at all ("packed file is corrupt") or crash immediately. In this case here the program runs perfectly fine but the *next* program you run crashes.

Completely different issues

**Comment by Akuma-Git** [comment]
Tested in SVN and STAGING, can't reproduce it, everything works fine.
Another reason to only support legitimate software, there is just too much crap out there.

**Comment by darkstar** [comment]
> Another reason to only support legitimate software, there is just too much crap out there.

@Akuma-Git  Wow. Nice accusation there... Thanks for implying that I'm playing pirated games. Well here is a surprise for you: my original copy of the game

![2020-09-06 21 26 12](https://user-images.githubusercontent.com/43795/92333721-0054d700-f088-11ea-9968-09366ba25d51.jpg)

Maybe YOUR copy of the game is crap?

**Comment by dreamer** [comment]
*Just before any conflict will escalate in comments… (because I don't want to moderate other user's posts - but will do it if the topic will stray away from discussing the issue).*

In the spirit of presuming good faith, I understood @Akuma-Git's comment as "we should not worry too much about supporting illegitimate/cracked copies", and not accusing anyone of owning one. But I see how it could be understood differently as well. I think it's just a small misunderstanding :)

Anyways, to continue constructively: @Akuma-Git, @darkstar : can you both post additional details of the environment you're testing in (OS, what game settings are being used?). And please remember to follow steps described in @darkstar's first post (corrupt chain is detected when running *next program* after Zone 66).

Side comment: when doing unrelated development, I recently found some hacks plastered with "zone66" comment (in CPU exception handling: [link](https://github.com/dosbox-staging/dosbox-staging/blob/master/src/cpu/cpu.cpp#L593)).

**Comment by kcgen** [comment]
This is a good point in general.. there might be different versions of a given game out there; and these kinds of bugs might be addressed in part, or new regressions might be introduced. 

I personally now have almost every Rayman 1.x and (regional version) in existence from kind interlopers who reported different CDDA and track offsets that have been incrementally fixed in subsequent versions.  (And I'll be the first to admit I don't own these.. I am simply trying to preserve a tiny bit of gaming history; and there are legal exclusions permitting this). 

There are also plenty of old and even modern examples of excessive copy protection fouling up the game itself or the user's PC, so in those cases the game is best experienced when 'liberated' from such garbage. 

None the less.. both those examples don't apply here (you're running a original, non-copy-protected instance), but wanted to mention that there are places for both cases to differ. 

Regarding possible version differences: is the version mentioned in a readme or shown on an intro-screen? If not, we can compare executable sizes (or hashes).   I think Zone 66 was released by an ex-demo scene folks out of Europe.. and there were lots of different level packs and addons; so they might have been tweaking it across these releases.

**Comment by darkstar** [comment]
My copy of Zone66 doesn't print a version number anywhere as far as I can see, so here's the SHA1's of the main programs:
````
  File: GAME.EXE
CRC-32: 29b2fcca
   MD4: 2ecdd15f2314c52b28c17750a2487989
   MD5: dc4588edcfdf1e21d08446abc81688be
 SHA-1: 03c8d6e18cebe7e25c07709c43bfde5c61b2672b

  File: ZONE66.EXE
CRC-32: 2d5b4804
   MD4: 40987ef2cda001ffe3a11376b46ef0c9
   MD5: 17ad2eb2e38ae24743d2b82a11fa6942
 SHA-1: 8a88cc020b874e2a6aab1b4d0dc115fe40cb119b
````

Problem still exists as of ce533d0 (downloaded the x64 build [from the CI](https://github.com/dosbox-staging/dosbox-staging/runs/1076009908) instead of compiling my own). I can also try on Linux


There are definitely multiple versions of the game, there is an "earlier" version (which used a very picky DOS extender and which wouldn't even start if HIMEM.SYS was installed, printing a message telling you to remove it and reboot) and a later one which swapped the DOS extender for something more mainline (I think it might have been DOS/32A).

The version I have is already the later version (from July 1993). I also have the earlier shareware version (from March 1993) which requires `xms off` and `ems false` (why one needs "off" and the other needs "false" is beyond me but hey ;-) ). And that one works just fine and the problem can't be reproduced there.

Yes, the earlier version with the "picky" DOS extender works fine, while the newer version craps all over the MCB chain.

This is the earlier version:
````
  File: GAME.EXE
CRC-32: 4a33087d
   MD4: 29bcd2c0c7209f659402473001a4666e
   MD5: f031c6916a2f88b5c5ee045f381f979e
 SHA-1: 51ff9c85a640877e89d518e4e7e4fe2be0be649c

  File: ZONE66.EXE
CRC-32: a2ada4b8
   MD4: ebed77fb5693ba02384e3f717d9a13bf
   MD5: bc9f4c35481846f10935560698d0a006
 SHA-1: 7965c7e37757ad7485c23d353d82b05b13a6a064
````

That gave me an idea: if I do `xms off` and `ems false` before starting the problematic Zone66 version, the problem actually disappears! I tried it a few times and in all but one cases I could play Zone66, quit, and then launch another game (Keen 4 in my case) without DosBox breaking.

There is still that one case where Zone66 itself crashed and took DosBox-staging down with it when starting a new game, but I couldn't reproduce that (to make a screenshot of the DosBox window, so it might have been a one-off). Still, some non-determinism seems to be involved here

**Comment by Akuma-Git** [comment]
> > Another reason to only support legitimate software, there is just too much crap out there.
> 
> @Akuma-Git Wow. Nice accusation there... Thanks for implying that I'm playing pirated games. Well here is a surprise for you: my original copy of the game
> 
> ![2020-09-06 21 26 12](https://user-images.githubusercontent.com/43795/92333721-0054d700-f088-11ea-9968-09366ba25d51.jpg)
> 
> Maybe YOUR copy of the game is crap? Any why should it have anything to do with windows that makes no sense...

Yeah, I rushed through it, sorry about that. Sometimes my assumptions go faster than the old brain. Its a race condition ;)

Well on Linux everything works fine, on both staging and the svn.
So that would leave open afaict:
1. Versioning, maybe older/newer had a bug/introduced a bug.
2. Windows specific, can be anything from programs, to drivers to updates icw latest build.
3. Media damage 'original' cd
4. Media damage 'image', (problems during creation, dust, wrong read..etc..)

**Comment by darkstar** [comment]
> 1. Versioning, maybe older/newer had a bug/introduced a bug.

Can you share details of what version of Zone66 you are running? Does it match one of the versions I posted above? Are you running with XMS/EMS on or off?

**Comment by kcgen** [comment]
Very nice round up @darkstar and discovering the least problematic dosbox configuration!

@Akuma-Git, I wonder if we're using different versions. Mine spews errors from the dosbox CPU core, however is perfectly playable. I've configured it for GUS.

My  md5sum:
``` text
zone66.exe 3ad5f717d2494059e2884f00af0fc615
```

I suspect I have 'not the first release', because the `helpme.doc` mentions:

```
Problem:  Zone 66 detects my Gravis Ultrasound card,  but I d on't get any music.

Solution: Please check your disk to make sure the file M03GMUZ.Z66 exists. If it doesn't,  then you have a copy of ZONE 66 which does not contain the Gravis soundtrack.  

To get the Gravis soundtrack, look for the file $ZONE66G.ZIP on your local BBS,  or you can get the latest version of ZONE 66 Mission 1 along with the Gravis sound track direct from Epic MegaGames for $5.  (See ORDERING INFO within the game)
```

**Comment by Akuma-Git** [comment]
Well its definitely different , as the version I used is came on floppy disks. Missions 1-4, 5-8
I will post it it the morning, I should be sleeping :yawning_face:

**Comment by darkstar** [comment]
Small info, `ZONE66.EXE` is just a loader, the actual game is in `GAME.EXE` so please share that file's checksums too.

Small tidbit: in some versions of the game you can skip the intro by running `GAME.EXE` directly, but you have to give it the command line arguments `Tran is Evil!!!`, i.e. `C:\zone66>GAME Tran is Evil!!!`  ;-)

This doesn't work in the earlier version though (the string is still there but it is slightly different and I didn't poke deep enough to find out what the correct arguments to GAME.EXE are in that version)

**Comment by darkstar** [comment]
Oh and one final comment about:
> Another reason to only support legitimate software, there is just too much crap out there.

I don't think dosbox-staging should have a "supported" and "not supported" list. This is primarily what drives me nuts with upstream DosBox, their insistence that it is "only for games and nothing else". If something runs in pure DOS, it should run in DosBox (or rather DosBox-staging). No matter if it's an original game, a pirated copy of a game, a database application or a control program for a nuclear reactor.... :-D

**Comment by kcgen** [comment]
Here's my game.exe details:

``` text
[02:31:18@hoya zone_66] md5sum game.exe
395fbeb1819c133cc314fceaf9d3a188  game.exe
[02:38:54@hoya zone_66] strings game.exe | grep -i tran
TranisEvil!!!L8r...$Critical file error... missing file!!!$Ran out of low memory!!!$Please run ZONE66.EXE$Checking for $Gravis Ultrasound.
666Tran666
```

**Comment by LowLevelMahn** [comment]
i've got 3 version of zone66 game.exe

318911529a5767aa225caa456bfd85db
395fbeb1819c133cc314fceaf9d3a188
dc4588edcfdf1e21d08446abc81688be

**Comment by darkstar** [comment]
> i've got 3 version of zone66 game.exe
>
> 318911529a5767aa225caa456bfd85db
> 395fbeb1819c133cc314fceaf9d3a188
> dc4588edcfdf1e21d08446abc81688be

but which ones exhibit the error and which don't? And with what memory configuration (EMS/XMS)?

> could also help to have some internal command available (like loadfix - checkmcb) to force a MCB chain check - so no other game is needed to detect the error or a config feature to always check mcb-chain after program end?

Ideally I think there should be a "debug mode" or option that one can enable that breaks into the debugger when some program overwrites critical DOS memory structures like that. Sure, it will slow everything down a bit, but for debugging this might still be useful

**Comment by Akuma-Git** [comment]
> Oh and one final comment about:
> 
> > Another reason to only support legitimate software, there is just too much crap out there.
> 
> I don't think dosbox-staging should have a "supported" and "not supported" list. This is primarily what drives me nuts with upstream DosBox, their insistence that it is "only for games and nothing else". If something runs in pure DOS, it should run in DosBox (or rather DosBox-staging). No matter if it's an original game, a pirated copy of a game, a database application or a control program for a nuclear reactor.... :-D

Yes it does, I do not waste my time on pirated versions because they have been patched the wrong way. That means, that I have to fix a problem someone else created first while there is a perfectly running legitimate copy. I'd rather use that time to help people like you that own their software.

On that note :smiley: , these are the check-sums from the 4x 1.44m disks release:

`
bfbb0dd79583d9c580800b6d67fa32cdafa8b39ac5758ba9da544892b49a48ed  GAME.EXE
dc90cacd1550198065ec0f797bb1580a0ba550b9d2d15af5d681f4ec2c76db41  HELPME.EXE
3a8393d8b8cbcbe44845a1a526c3ecfbc391fdf8cd07b0a834829cda5a0b9edb  INTRO.EXE
8cf0939025f6656c714f663a1ebf60061ada2c46bf8dcb9c49b2413b46cb58e9  ZONE66.EXE
`

**Comment by LowLevelMahn** [comment]
> > i've got 3 version of zone66 game.exe
> > 318911529a5767aa225caa456bfd85db
> > 395fbeb1819c133cc314fceaf9d3a188
> > dc4588edcfdf1e21d08446abc81688be
> 
> but which ones exhibit the error and which don't? And with what memory configuration (EMS/XMS)?

didn't test it, just want to mention the version i got - if there is further testing needed

> 
> > could also help to have some internal command available (like loadfix - checkmcb) to force a MCB chain check - so no other game is needed to detect the error or a config feature to always check mcb-chain after program end?
> 
> Ideally I think there should be a "debug mode" or option that one can enable that breaks into the debugger when some program overwrites critical DOS memory structures like that. Sure, it will slow everything down a bit, but for debugging this might still be useful

aren't there progams that tend to write into this structures on purpose? i don't think games do it but who knows

**Comment by darkstar** [comment]
> I do not waste my time on pirated versions because they have been patched the wrong way. That means, that I have to fix a problem someone else created first while there is a perfectly running legitimate copy. I'd rather use that time to help people like you that own their software.

But if something runs on DOS but not on DosBox, this shows that DosBox is not perfect and that there's a bug somewhere. This bug should be solved instead of saying "nah, who cares, it only happens with DBase3 so far and we don't want people running that on DosBox anyway", like the mainline devs do.

And don't worry, nobody will force you to help with bug report X if you suspect that it only happens with pirated software. So yeah, you can still decide to not help in that case, without needing an official "supported/unsupported" list.

>> Ideally I think there should be a "debug mode" or option that one can enable that breaks into the debugger when some program overwrites critical DOS memory structures like that. Sure, it will slow everything down a bit, but for debugging this might still be useful

> aren't there progams that tend to write into this structures on purpose? i don't think games do it but who knows

I'm not sure, I thought the MCB chain was completely under the control of the DOS kernel. Still, doing a before/after sanity check of the MCB structure on a rewrite is entirely possible

**Comment by LowLevelMahn** [comment]
> But if something runs on DOS but not on DosBox, this shows that DosBox is not perfect and that there's a bug somewhere. This bug should be solved instead of saying "nah, who cares, it only happens with DBase3 so far and we don't want people running that on DosBox anyway", like the mainline devs do.

the question is still: is the MCB chain also damaged under real DOS but undetected due to missing checks
is there something like an pure DOS MCB check tool? (source code here: https://www.vogons.org/viewtopic.php?f=31&t=43440)

**Comment by kcgen** [comment]
Brilliant find @LowLevelMahn ; we can wire that up to a hotkey to run a scan on demand (and make that available in release builds), and going going further it would be nice to have a compile-time memcheck flag that adds valgrind-type canary and check code around memory read and write calls so these voilations are flagged in realtime.

@darkstar, I also agree that we ideally want to provide dos+hardware-equivalent behavior for any goven snippet of code, when it doesn't harm the projects goals in general. (Ie: introducing new bugs or regressions, significantly reducing performance, or complexifying the code making maintenance more difficult).

**Comment by Akuma-Git** [comment]
> But if something runs on DOS but not on DosBox, this shows that DosBox is not perfect and that there's a bug somewhere. This bug should be solved instead of saying "nah, who cares, it only happens with DBase3 so far and we don't want people running that on DosBox anyway", like the mainline devs do.

I think if companies step up through actively contributing or donations that it would be different. Now they can easily run it in a VM. Unless they come with a detailed bug report, I personally do not care.

> And don't worry, nobody will force you to help with bug report X if you suspect that it only happens with pirated software. So yeah, you can still decide to not help in that case, without needing an official "supported/unsupported" list.

I'm not.

On topic: Where can I send you a link for some testing ?

**Comment by LowLevelMahn** [comment]
@darkstar 
>But if something runs on DOS but not on DosBox, this shows that DosBox is not perfect and that there's a bug somewhere.

did you check under DOS or FreeDOS? 

FreeDOS seems to have a buildin MCB chain validation like dosbox got (https://www.mail-archive.com/freedos-user@lists.sourceforge.net/msg04201.html showing a "PANIC: MCB chain corrupted" message)

**Comment by kcgen** [comment]
[mm1_mcbfix.zip](https://github.com/dosbox-staging/dosbox-staging/files/5187132/mm1_mcbfix.zip)

**Comment by kcgen** [comment]
As ripsaw8080 [explains](https://www.vogons.org/viewtopic.php?p=422433#p422433), DOSBox perform MCB chain consistency checks and reports them as seen above.

The cause for the corrupt MCB chain is due to the emulated program scribbling over its memory (which DOSBox catches, unless real hardware which was liable to lockup or crash). 

Closing this, as the issue lies outside of DOSBox.
